File: leetcode/Easy/0190-reverse-bits/0190-reverse-bits.cpp
/*
 Problem: 190. Reverse Bits
 Language: cpp
 Runtime: 0 ms (100.00%)
 Memory: 7.6 MB (100.00%)
 Tags: Divide and Conquer, Bit Manipulation
*/
class Solution {
public:
    uint32_t reverseBits(uint32_t n) {
        uint32_t result = 0;
        for (int i = 0; i < 32; ++i) {
            result <<= 1;          // Shift the result to the left by 1 bit
            result |= (n & 1);     // Add the least significant bit of n to the result
            n >>= 1;               // Shift n to the right by 1 bit
        }
        return result;
    }
};

File: leetcode/Easy/0190-reverse-bits/0190-reverse-bits.rs
/*
 Problem: 190. Reverse Bits
 Language: rust
 Runtime: 2 ms (52.47%)
 Memory: 2.1 MB (92.52%)
 Tags: Divide and Conquer, Bit Manipulation
*/
impl Solution {
    pub fn reverse_bits(n: i32) -> i32 {
        let mut v: i32 = 0;
        let mut x = n;

        for _ in 0..32 {
            v = (v << 1) | (x & 1);
            x >>= 1;
        }

        v   
    }
}